### PR TITLE
fix(core): terminal handling for content-filter refusals with auto server-side fallback to opus for fable-5

### LINKED ISSUE #17888: Mastra Code v0.22.3 with fable-5 can silently stop without a terminal state
## Description

When using Mastra Code `v0.22.3` with the `fable-5` model, an active run can intermittently stop making progress without being marked as completed, failed, terminated, cancelled, or timed out.

From the user perspective, execution just stops mid-task:
- no final assistant message
- no explicit error
- no `terminated` / `failed` / `cancelled` / `timeout` state
- no clear indication whether the run is still active or dead

This has happened during longer multi-step coding sessions involving tool calls, file edits, tests, commits, and verification work.

## Expected behavior

Every Mastra Code run should eventually emit a clear terminal state, such as:

- `completed`
- `failed`
- `terminated`
- `cancelled`
- `timeout`

If the model stream, tool runner, orchestration layer, or CLI connection fails, the user should see an explicit failure/timeout message instead of the run silently stopping.

Ideally the message should include enough diagnostic context to debug it, for example:
- run/session ID
- model request ID, if available
- last emitted assistant/tool event
- whether the model stream closed normally
- whether a tool call was active
- whether a timeout/watchdog fired

## Actual behavior

The run sometimes stops without reaching any visible terminal state.

The session is left ambiguous:
- it is not clearly active
- it is not clearly completed
- it is not clearly failed
- the user has to manually inspect/retry/continue the work

## Steps to reproduce

This is intermittent, but the observed pattern is:

1. Start Mastra Code `v0.22.3`.
2. Use model `fable-5`.
3. Run a longer coding task that uses multiple tool calls, for example:
   - inspect files
   - edit code
   - run tests/builds
   - commit changes
   - verify deployment or repo state
4. Let the agent run for a while.
5. Sometimes execution stops mid-flow without any terminal state or final error message.

## Impact

This makes long-running coding tasks unreliable because the user cannot tell whether the run:
- is still running
- got stuck
- crashed internally
- lost connection
- timed out
- was cancelled

It also makes autonomous/goal-style workflows difficult to trust, because completion cannot be determined reliably.

## Environment

- Product: Mastra Code
- Mastra Code version: `v0.22.3`
- Model: `fable-5`
- Interface: CLI / interactive coding agent
- OS: Darwin
- Node.js: `v24.16.0`

## Duplicate check

I searched open issues with:
- `mastracode` label
- `fable-5`
- `mastracode stops`
- `mastracode stuck`
- `terminal state`

I found one partially related issue mentioning runs not reaching a terminal state, but it appears to be about `@mastra/core` Observational Memory / background task queue behavior rather than Mastra Code + `fable-5`, so this does not look like a direct duplicate.

## Requested fix

Please ensure Mastra Code always finalizes a stopped run into an explicit terminal state.

Suggested behavior:
1. Add a watchdog/finalizer for runs that stop emitting events.
2. Convert silent model/tool/orchestration failures into `failed` or `timeout`.
3. Surface a clear user-facing message in the CLI.
4. Include diagnostic IDs/events where possible.
5. Add logging around model stream closure and tool-call lifecycle transitions.

## Acceptance criteria

- A Mastra Code run using `fable-5` cannot silently stop without a visible terminal state.
- If the run fails or times out, the CLI shows an explicit failure/timeout message.
- The user can reliably distinguish active, completed, failed, cancelled, and timed-out runs.

### PR DESCRIPTION
**Pull request by abhiaiyer91**
## Summary

Fixes #17888 — runs with `claude-fable-5` could hang forever or stop silently with no terminal state when the model's safety classifier blocked a turn.

Anthropic classifier blocks surface through the AI SDK as a `content-filter` finish reason (vercel/ai#15929). Two independent bugs produced the reported symptom, and this PR fixes both plus adds resilience and observability around server-side fallbacks:

### 1. Loop: treat `content-filter` as terminal (the hang)
The agentic loop's continuation logic only excluded `stop`/`error`/`length`, so a refusal re-sent the identical request, re-triggered the identical refusal, and spun until `maxSteps` (1000 in the Harness). A new `TERMINAL_FINISH_REASONS` constant now includes `content-filter`, and the pending-tool-call guard can no longer override `error`, `length`, or `content-filter` back into continuing.

### 2. Harness: surface non-success finish reasons (the silent stop)
The harness mapped every non-tool finish reason to `stopReason: 'complete'`, so a refused run looked like a quiet success: empty message, no error, no terminal state. The finish handler now maps `content-filter`/`error`/`length` to `stopReason: 'error'` with a user-facing message (extracting Anthropic refusal category/explanation from `stopDetails`, handling both finish-chunk payload shapes), and both stream finalization paths emit an `error` event and end the run with `reason: 'error'`.

### 3. Harness: auto server-side fallback for `claude-fable-5`
When the active model is `claude-fable-5`, the harness injects `providerOptions.anthropic.fallbacks: [{ model: 'claude-opus-4-8' }]` so a classifier-blocked turn is transparently retried server-side instead of refusing. The AI SDK adds the required beta header automatically. If the whole chain refuses, the terminal handling above kicks in.

### 4. Fallback visibility
When a fallback actually serves a turn (reported via `fallback_message` entries in `providerMetadata.anthropic.iterations`), the harness emits an `info` event so the UI can tell the user which model answered.

### 5. Tracing attribution
The `MODEL_GENERATION` span's `responseModel` attribute now reports the fallback model when one served the turn, so tracing exporters (Langfuse, etc.) attribute usage/cost to the model that actually generated the response. Detection logic is shared via a new `server-side-fallback` module used by both the harness and `model.loop.ts`.

## Test plan

- New colocated tests (all pass):
  - `harness-finish-reason.test.ts` (9): content-filter with/without provider stop details, length, normal-stop baseline; message-builder unit tests
  - `harness-fable-fallback.test.ts` (9): fallback injection across model-id formats, negative cases, fallback notice building
  - `server-side-fallback.test.ts` (9): fallback detection and response-model resolution, incl. malformed metadata
  - `llm-execution-step.test.ts` (+3): content-filter and error finish reasons are terminal even with pending tool calls
- Full suites: 345 harness + 363 llm/model tests pass, typecheck and lint clean
- Verified live: a prompt that previously hung forever now terminates promptly with "Content was filtered by the model's safety system."

Changeset: `@mastra/core` patch.

## ELI5 Explanation

When using Claude Fable-5, runs would sometimes hang or stop without warning if the model refused to answer (content-filter). This PR fixes it by: (1) treating refusals as final instead of trying again indefinitely, (2) showing clear error messages when this happens, and (3) automatically falling back to Claude Opus if available so the task can continue instead of failing.

## Problem Statement

Runs using `claude-fable-5` intermittently stopped emitting events without reaching a terminal state—no final message, no explicit error, and no completed/failed/terminated/cancelled/timeout status. When Anthropic's safety classifier returned a `content-filter` finish reason, the agent loop would hang or silently stall, leaving sessions ambiguous and making autonomous workflows unreliable.

### Terminal State Handling

- Introduced a centralized `TERMINAL_FINISH_REASONS` constant (`stop`, `error`, `length`, `content-filter`) in the agent execution loop so refusals are treated as terminal and do not trigger re-sends or pending tool call retries.
- Updated `describeNonSuccessFinishReason()` helper to convert non-success finish reasons (especially `content-filter`) into user-facing error text by extracting Anthropic refusal category and explanation from `providerMetadata.anthropic.stopDetails`.
- Harness now maps `content-filter`, `error`, and `length` to `stopReason: 'error'` with populated `errorMessage` fields and emits `error` events with diagnostic context.
- Runs ending on these non-success reasons finalize with `agent_end` reason of `'error'` instead of normal completion.

### Server-Side Fallback for Fable-5

- Added `buildFableFallbackProviderOptions()` to detect `claude-fable-5` model IDs and inject `providerOptions.anthropic.fallbacks: [{ model: 'claude-opus-4-8' }]` into harness stream options.
- When a fallback chain still refuses, the same terminal handling applies.
- Fallback injection uses shallow merge logic; note that existing caller-provided `providerOptions.anthropic.fallbacks` may be overwritten (reviewer flagged potential issue for concatenation instead).

### Fallback Visibility and Tracing

- Emits an `info` event naming the actual model when a server-side fallback serves a turn, detected via `describeServerSideFallback()` which inspects `providerMetadata.anthropic.iterations.fallback_message`.
- Added `server-side-fallback.ts` module with `getServerSideFallbackInfo()` and `resolveResponseModelId()` utilities.
- MODEL_GENERATION span's `responseModel` attribute now reports the actual fallback model when one served the turn (instead of just the declared model ID).

## Test Coverage

- `harness-finish-reason.test.ts`: Unit and end-to-end tests for `describeNonSuccessFinishReason()` and harness finalization behavior (content-filter with Anthropic refusal details, length finish, fallback-served stops, plain stop).
- `harness-fable-fallback.test.ts`: Tests for `buildFableFallbackProviderOptions()` across three model ID formats and edge cases, plus `describeServerSideFallback()` detection.
- `server-side-fallback.test.ts`: Tests for `getServerSideFallbackInfo()` and `resolveResponseModelId()` detection and fallback.
- `llm-execution-step.test.ts`: Tests verifying content-filter and error finish reasons are terminal even with pending tool calls.
- Full harness and llm/model test suites pass (345 harness + 363 llm/model).

## Notes

- Changeset documents patch for `@mastra/core`.
- Reviewer flagged that hardcoded model names (`claude-fable-5`, `claude-opus-4-8`) in changeset text and tests should use registered tokens from `docs/src/plugins/remark-model-tokens/models.ts` per repo policy (runtime logic may keep literals).
- Minor regex consideration: the suffix match `(^|\/)claude-fable-5$` could match non-Anthropic-routed IDs (e.g., `openrouter/anthropic/claude-fable-5`), leading to Anthropic options being written for models a downstream provider ignores.
- Core loop centralization and terminal-state surfacing verified correct; local tests and typechecking pass.

### CODE REVIEW

**Comment by CalebBarnes** [comment]
Two suggestions before this goes in:

**1) AGENTS.md model-token rule.** The repo guideline is to use placeholder tokens from `docs/src/plugins/remark-model-tokens/models.ts` whenever a model name/ID appears in changesets, tests, or comments. The changeset narrative, the new JSDoc in `harness.ts`/`llm-execution-step.ts`, and both new test files hardcode `claude-fable-5` and `claude-opus-4-8`. The runtime regex/branch obviously has to keep the literal string, but the user-facing changeset text and the comments/tests should probably use tokens (and `models.ts` may need new entries for these IDs).

**2) `buildFableFallbackProviderOptions` shallow-merges fallbacks.** In `buildStreamOptions`, the merge is:
```ts
anthropic: { ...existingAnthropic, ...fableFallback.anthropic }
```
If a caller already passed `providerOptions.anthropic.fallbacks: [...]`, we silently overwrite their list with `[{ model: 'claude-opus-4-8' }]`. Worth either (a) skipping injection when `existingAnthropic.fallbacks` is set, or (b) concatenating arrays.

Minor: the suffix regex `(^|\/)claude-fable-5$` matches non-Anthropic-routed IDs like `openrouter/anthropic/claude-fable-5`. For those we still write `anthropic` providerOptions, which the downstream provider will likely ignore — but it's a small footgun worth noting.

The core loop fix (centralizing `TERMINAL_FINISH_REASONS` and making `content-filter` terminal even with pending tool calls) and the harness terminal-state surfacing both look correct. All targeted tests + the full `harness` and `agentic-execution` suites pass locally; typecheck is clean.

File: packages/core/src/loop/workflows/agentic-execution/llm-execution-step.ts [hunk 1]
```diff
@@ -2009,10 +2024,20 @@ export function createLLMExecutionStep<TOOLS extends ToolSet = ToolSet, OUTPUT =
...
+      // finishReason='stop' alongside tool calls, which the loop must process.
+      const hasPendingToolCalls =
+        toolCalls &&
+        toolCalls.some(tc => !tc.providerExecuted) &&
+        finishReason !== 'length' &&
+        finishReason !== 'content-filter';
       const shouldContinue =
         shouldRetry ||
-        (!tripwireTriggered && (hasPendingToolCalls || !['stop', 'error', 'length'].includes(finishReason)));
+        (!tripwireTriggered && (hasPendingToolCalls || !TERMINAL_FINISH_REASONS.includes(finishReason)));
```
[the remark this answers is not in the document]
Comment (abhiaiyer91) [reply packages/core/src/loop/workflows/agentic-execution/llm-execution-step.ts]: Fixed in adeb641420 — `hasPendingToolCalls` now also excludes `error`, with a regression test (error finish + pending tool call → `isContinued: false`).

File: packages/core/src/harness/harness.ts [hunk 2]
```diff
@@ -2820,7 +2918,19 @@ export class Harness<TState = {}> {
...
+          // the run ends with no final message and no error, leaving the user
+          // unable to tell whether it completed, failed, or is still active.
+          const errorMessage = describeNonSuccessFinishReason(finishReason, chunk.payload?.providerMetadata);
+          if (errorMessage) {
+            state.currentMessage.stopReason = 'error';
+            state.currentMessage.errorMessage = errorMessage;
+            state.terminalError = errorMessage;
+          } else {
+            state.currentMessage.stopReason = 'complete';
+          }
```
[the remark this answers is not in the document]
Comment (abhiaiyer91) [reply packages/core/src/harness/harness.ts:2971]: Fixed in adeb641420 — the call site now reads `chunk.payload?.metadata?.providerMetadata ?? chunk.payload?.providerMetadata`, matching the loop's normalization.

File: packages/core/src/harness/harness.ts [hunk 3]
```diff
@@ -81,6 +87,66 @@ type HarnessSendNotificationSignalOptions = {
...
+ *
+ * The match is suffix-based so it covers `anthropic/claude-fable-5`, a bare
+ * `claude-fable-5`, and any pack/provider-prefixed form.
+ */
+export function buildFableFallbackProviderOptions(modelId: string): { anthropic: { fallbacks: { model: string }[] } } | undefined {
+  if (!/(^|\/)claude-fable-5$/.test(modelId)) {
+    return undefined;
+  }
+  return { anthropic: { fallbacks: [{ model: FABLE_FALLBACK_MODEL }] } };
+}
```
[the remark this answers is not in the document]
Comment (abhiaiyer91) [reply packages/core/src/harness/harness.ts:149]: Skipping this one: these comments document literal runtime values — the harness's suffix regex specifically matches `claude-fable-5` and the hard-coded fallback constant is `claude-opus-4-8`. No placeholder token exists for either model in `docs/src/plugins/remark-model-tokens/models.ts`, and substituting a rotating token would misdocument the actual behavior. The placeholder rule targets examples/docs where model versions rotate, not constants describing pinned runtime behavior.

File: packages/core/src/loop/workflows/agentic-execution/llm-execution-step.ts [hunk 4]
```diff
@@ -55,6 +55,21 @@ import type { CollectedChunk } from './build-messages-from-chunks';
...
+ * (or forever when maxSteps is unset).
+ *
+ * - `stop`: the model finished normally.
+ * - `error`: the model stream failed.
+ * - `length`: the model hit max_tokens; retrying reproduces the truncation
+ *   (issue #15717).
+ * - `content-filter`: a classifier block / model refusal (e.g. `claude-fable-5`
+ *   surfaced by the AI SDK as `content-filter`). Retrying re-triggers the same
+ *   refusal, so the run would hang indefinitely.
+ */
```
[the remark this answers is not in the document]
Comment (abhiaiyer91) [reply packages/core/src/loop/workflows/agentic-execution/llm-execution-step.ts:70]: Skipping for the same reason as the harness comment: `content-filter` here is specifically about fable-5's classifier behavior baked into the runtime logic, and no placeholder token exists for `claude-fable-5`.

**Review by CalebBarnes — APPROVED** [review]
LGTM. Both CodeRabbit findings (terminal `error` with pending tool calls, nested `providerMetadata` shape) addressed correctly, and the new server-side fallback notice cleanly surfaces the silent-substitution case. Tests pass locally (39 in changed files, full harness + agentic-execution suites green), types clean.

File: packages/core/src/llm/model/server-side-fallback.test.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      getServerSideFallbackInfo({
+        anthropic: {
+          iterations: [
+            { type: 'message', model: 'claude-fable-5', inputTokens: 10, outputTokens: 0 },
+            { type: 'fallback_message', model: 'claude-opus-4-8', inputTokens: 10, outputTokens: 50 },
+          ],
+        },
+      }),
+    ).toEqual({ model: 'claude-opus-4-8' });
+  });
```
[the remark this answers is not in the document]
Comment (abhiaiyer91) [reply packages/core/src/llm/model/server-side-fallback.test.ts]: Fixed in 6ff4678309 — replaced `claude-opus-4-8` with the registered token value `claude-opus-4-7` in both fixture blocks (the fixture value is arbitrary for the detection logic). Kept `claude-fable-5` per the repo exception, and left the `harness-fable-fallback.test.ts` assertions unchanged since those pin the actual runtime constant `FABLE_FALLBACK_MODEL`.
